---
title: "Workflow of GMA"
output: html_pdf
date: "07-21-2022"
author: "Zhikai Yang, Qi Zhang, Jinliang Yang"
---


## Installation

- __Running environment__: 
    - The workflow was constructed based on the __Linux system__ running the [R v4.1.1](https://cran.r-project.org/).

- __Required software and versions__: 
    - [Rstudio](https://www.rstudio.com/products/rstudio/download/)
      
```{r, eval=FALSE}
install.packages(c("data.table", "glmnet", "MASS", "rrBLUP", "parallel", 
                   "doParallel", "CompQuadForm", "circlize", "dplyr"))
```

## Input Data

#### Required input data:

- __y matrix__ (`input/y_matrix.txt`): 
  - A `n x 1` matrix of phenotype values; `n` is the number of individuals.
  
```{r}
library("data.table")
y <- fread("input/y_matrix.txt", header=T,data.table=FALSE)
y = as.matrix(y)
dim(y) #271   1
head(y)
```

  
- __Z matrix__ (`input/Z_matrix.txt`): 
  - The `n x m` genotype matrix of data; `m` is the number of bi-allelic SNP markers, coded as `-1, 0, 1`.

```{r}
Z <- fread("input/Z_matrix.txt", header=T,data.table=FALSE)
Z = as.matrix(Z)
dim(Z) #271 5000
Z[1:10, 1:10]
```
  
- __X matrix__ (`input/X_matrix.txt`): 
  - The `n x k` intermediate Omics matrix; `k` is the number of Omics traits. 
  In the example, gene expression data (i.e., RNA-seq read counts of `k` genes) was used as the intermediate traits .

```{r}
X <- fread("input/X_matrix.txt", header=T,data.table=FALSE)
X = as.matrix(X)
dim(X) #271 1200
X[1:10, 1:10]
```


#### Optional input data:

- __X0 matrix__ (`input/X0_matrix.txt`): 
  - A matrix of confounding effects. In the example, three principal components calculated from the Z matrix were used to control population structure.
  
```{r}
X0 <- fread("input/X0_matrix.txt", header=T,data.table=FALSE)
X0 = as.matrix(X0)
dim(X0) #271   3
head(X0)
# X0 = prcomp(Z)$x[,1:3] # use this line of code to calculate principal components if no X0_matrix.txt file provided
```

## Major steps

#### Step 1: Calculate confounding effect (or the `X0` matrix)

Several principal components can be used as the fixed effects to control population structure as the confounding effects if using a structured population.

```{r}
source('lib/utils.R')
        
Z <- fread("input/Z_matrix.txt", header=T, data.table=FALSE)
Z = as.matrix(Z)
X0 <- getpca(Z, p=3) # here the first p=3 PCs were extracted.
```

#### Step 2: Conduct GMA using different methods


```{r}
library(data.table)
#library(GMA)
library(glmnet)
library(MASS)
library(rrBLUP)
library(parallel)
library(doParallel)
library(CompQuadForm)
source('lib/highmed2019.r')
source('lib/fromSKAT.R')
source('lib/MedWrapper.R')
source('lib/reporters.R')

subX = X[, 1:100]
# run the fixed effect model that assign equal penalty on the two data types.
run_GMA(y, X0, subX, Z, ncores=10, model="MedFix_eq", output_folder="output/")

# run the fixed effect model that minimizes BIC
run_GMA(y, X0, subX, Z, ncores=10, model="MedFix_fixed", output_folder="output/")

# run the random effect model using linear kernel, and extract the model that minimizes BIC
run_GMA(y, X0, subX, Z, ncores=10, model="MedMix_linear", output_folder="output/")

# run the random effect model using shrink_EJ kernel, and extract the model that minimizes BIC
run_GMA(y, X0, subX, Z, ncores=10, model="MedMix_shrink", output_folder="output/")

```

#### Step 3: Visualize the results

- You can plot the results yourself using the below R code.

```{r}
library(circlize)
library(dplyr)

gwas <- qGWAS(y, Z, plot=FALSE)
fwrite(gwas, "output/gwas_results.csv", sep=",", row.names = FALSE, quote=FALSE)



circos_med(gwas_res="output/gwas_results.csv",
           med_res="output/mediators_fixed_bic_trait_V1.csv", 
           dsnp_res="output/dsnps_fixed_bic_trait_V1.csv", 
           isnp_res="output/isnps_fixed_bic_trait_V1.csv",
           chrlen="input/Chromosome_v4.txt", 
           gene_position= "input/gene_pos.csv",
           out_tiff = "graphs/circos.tiff")
```

## Expected results

The outputs of the example data:  

#### dSNP: 
- `output/dSNP_MedFix_eq_trait_V1.csv` and `output/dSNP_MedFix_fixed_trait_V1.csv`: direct SNPs identified using `MedFix_eq` and `MedFix_fixed` methods for trait `V1`. Note that only MedFix methods will report direct SNP. The output file contains the following columns: 
  - snp: direct SNP; 
  - pval: p-value of effect from exposure to outcome; 
  - coef: SNP effect from exposure to outcome.

#### iSNP:
- `output/iSNP_MedFix_eq_trait_V1.csv` and `output/iSNP_MedFix_eq_trait_V1.csv`: indirect SNPs identified. Again, only `MedFix` methods will report direct SNPs. The output file contains the following columns: 
  - medi: mediator gene under control by the iSNP;
  - snps_for_medi: indirect SNPs for the corresponding mediator; 
  - coef: effect from exposure to mediator.

#### Mediator:
- `output/mediator_MedFix_eq_trait_V1.csv`, `output/mediator_MedFix_eq_trait_V1.csv`, `output/mediator_MedMix_linear_trait_V1.csv`, `output/mediator_MedFix_eq_trait_V1.csv`: non-adjusted mediators detected by different methods of `MedFix_BIC`, `MedFix_0.5`, `MedMixed_Linear`, and `MedMixed_Shrink` for trait V1.
  - id: mediator gene id; 
  - e2m: p-value of effect from exposure to mediator; 
  - m2y: p-value of effect from mediator to outcome; 
  - e2m2y: maximum value between e2m and m2y; 
  - padj: adjusted p-value; 
  - coef: product of effect from exposure to mediator and effect from mediator to outcome.




### Visualization of GMA results:

![](graphs/circos.PNG)

- In the circos plots, the outermost circular track represents the ten chromosomes; 
- The next inner track shows the GWAS results, with two circular blue dashed lines indicating -log(p-value) of 5 and 10 and the red lines denoting the position of direct SNPs;
- The next inner track shows the relative positions of identified mediator genes with different genes represented by different colors; the lines in the innermost circle connects mediators with their corresponding indirect SNPs.
